keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1347 @@
+example: let's assume we want to fail a job if the container exit code
+(for "main-container") is in range [1,10] or [100,127] . Also, we want
+to ignore the pod failure if the pod condition type is "Preempted" or "Evicted".
+
+In the current API we can formulate this policy as follows:
+
+```yaml
+ backoffPolicy:
+   rules:
+   - action: Terminate
+     requirements:
+     - onExitCodes:
+         containerName: main-container
+         inRange:
+         - minInclusive: 1
+           maxInclusive: 10
+         - minInclusive: 100
+           maxInclusive: 127
+   - action: Ignore
+     requirements:
+     - onPodCondition:
+       - key: Status
+         operator: In
+         values:
+         - True
+       - key: Type
+         operator: In
+         values:
+         - Preempted
+         - Evicted
+```
+
+Alternative: Source-based
+
+This approach unifies the API for exit codes and pod conditions using the
+concept of an operator:
+
+```yaml
+ backoffPolicy:
+   rules:
+   - action: Terminate
+     requirements:
+     - source: ExitCode
+       containerName: main-container
+       operator: InClosedRange
+       values:
+       - 1
+         10
+     - source: ExitCode
+       containerName: main-container
+       operator: InClosedRange
+       values:
+       - 100
+         127
+   - action: Ignore
+     requirements:
+     - source: PodConditionType
+       operator: AnyIn
+       values:
+       - Preempted
+       - Evicted
+```
+
+Pros:
+- unifies the approach for handling exit codes and pod statuses with operators
+
+Cons:
+- it is more verbose that the main approach
+- these operators are non-standard as they don't use the `key` field (since
+  they are not used against maps as the label selector operators are)
+- requires introduction of a non-standard operator `InClosedRange` to shorten
+  the syntax (compared to using `Lt` and `Gt` operators). The new operator(s)
+  can be confusing for users used to label selector operators
+- there is a list of conditions (and thus condition types), so the operator
+  `In` would be misleading. The operator `AnyIn` would better reflect the
+  intention, but it would be another non-standard operator
+- it is hard to achieve the flexibility which we can have in the current
+  approach. For example, it is not clear how to (1) exclude condition types
+  corresponding to conditions with status=False, or how to check presence of
+  conditions with Type=Evicted and Reason=NodePressure.
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Evolving condition types
+
+The list of available PodCondition types field will be evolving with new
+values being added and potentially some values becoming obsolete. This can make
```

Comment:
it also makes status a contract.  Like this rename:
https://github.com/kubernetes/kubernetes/issues/102820 will be very problematic when we have this kind of dependency.

Reply:
But that's not a problem with ConditionTypes, as they are limited as part of the API https://github.com/kubernetes/api/blob/edebc6706ddef7e828b5b606fbd0c91c6843c2fa/core/v1/types.go#L2646-L2656

I think this is another reason not to filter by `reason` :)

Reply:
Agree, for that matter, we should encourage only the use of `type` in the API and discourage the use of `reason`